### DISCUSSION TITLE
The tool modify_fates_paramfile.py could change the variable have other dimension?

### DISCUSSION THREAD

**Discussion by LianGong98** [NGEET/fates#1015] [Q&A] [answered]

**Comment by rgknox** [comment, marked as answer]
Hi @LianGong98 , This is a great question and might help other users, so I'll re-open.  Sorry for the delay.  Below is a patch file I created to demonstrate how I would do this.   

In the example, I want to change the hydraulic vulnerability parameter "fates_hydro_avuln_node".  This is a 2d array, with dimensions PFT x hydraulic-organ.  There are 4 hydraulic organs, and they are leaves, stems, transporting roots and absorbing roots (in that order).  Note that this parameter, and every other 2D parameter in the parameter file, assumes that the PFT dimension is the outer dimension. 

I create a "patch" file, that will direct a python script to read in a base CDL file "ie base_file", make the changes that are specified, and generate the results in a new file "ie new_file".  

In the first manipulation, it will get rid of the PFTs in the base file that the user isn't interested in. In this example, I suppose we are not interested in anything but tropical PFTs, so set it to only keep the "broadleaf_evergreen_tropical tree", the "broadleaf_hydrodecid_tropical_tree" and the "c4_grass".  After it culls the rest, they are now considered PFT 1, 2 and 3 (in that order).

In the second manipulation, it will update the vulnerability parameters (for all four organs), but I tell it to only make changes to the "broadleaf_evergreen_tropical _tree" and the "c4_grass" (which are now PFT 1 and 3 respectively).  The comments in the script explain the convention used.

To execute this script, you must have python installed, and then change to the directory "fates/parameter_files/".  Issue the command:

> ../tools/BatchPatchParams.py --f=patch_hydro_example.xml

Here is the contents of the file "patch_hydro_example.xml":

```
<?xml version="1.0"?>
<all>
  <notes>This is a demonstration of changing the values in an array that is more than
  the PFT dimension. These values and selections have no real meaning. </notes>
  <base_file>fates_params_default.cdl</base_file>
  <new_file>fates_example_avuln_hydro.cdl</new_file>

  <!-- Lets say we are interested in only simulating tropical PFTs
       broadleaf_evergreen_tropical_tree    old pft # = 1, new pft # = 1
       broadleaf_hydrodecid_tropical_tree   old pft # = 5, new pft # = 2
       c4_grass                             old pft # = 12, new pft # = 3   -->
  
  <pft_trim_list>1,5,12</pft_trim_list>

  <parameters>
    
    <!-- Lets assume we have vulnerability data for only the evergreen
	 and the grass, I don't know why, its just an example 
	 so we select pft 1 and 3, and leave the hydrodeciduous with the default-->
    
    <pft_group ids="1,3">

      <!-- The default vulnterability param is 2 for all organs
	   Organ indexing is leaf, stem, transporting root, absorbing root
	   For the evergreen: leaves = 1, stem = 2, transporting root = 2, absorbing root = 3
	   For the grass:     leaves = 2, stem = 3, transporting root = 3, absorbing root = 4
	   These are probably nonsense, just demonstrating how to fill things in -->
      
      <fates_hydro_avuln_node> 1, 2, 2, 3,
                               2, 3, 3, 4 </fates_hydro_avuln_node>
    </pft_group>
  </parameters>
</all>


```

Alternatively, you could call modify_fates_paramfile.py directly to make change. This script works on the binary form of the parameter file (ie the .nc  extension) and not the text version (ie. the cdl extension).

Lets say that I want to now change the vulnerability of all four organs, for the  "broadleaf_hydrodecid_tropical_tree" from the file I generated above, from 2 to 1.

First, I would convert the result to the binary form:

> ncgen -o fates_example_avuln_hydro.nc fates_example_avuln_hydro.cdl

Then I would call the python script:

> ../tools/modify_fates_paramfile.py --fin fates_example_avuln_hydro.nc --fout fates_example_avuln_hydro_v2.nc --pft 2 --var fates_hydro_avuln_node --val 1,1,1,1

If I dump out the file and look at the vulnerability, it should look like this:

> ncdump -v fates_hydro_avuln_node fates_example_avuln_hydro_v2.nc

```
 fates_hydro_avuln_node =
  1, 1, 2,
  2, 1, 3,
  2, 1, 3,
  3, 1, 4 ;

```

**Reply by LianGong98** [reply]
@rgknox Thanks so much!  It helps me a lot !
  I have another question.When the plant hydro turned on, the canopy transpiration is always low, I try to modify some parameters ,shuch as _fates_hydro_avuln_gs ,fates_hydro_p50_gs ,fates_hydro_kamx_node_... but no obvious effects.
  
If I want to make the value of canopy transpiration larger,what should I do? 
Here is the output from the first year, but most of  value should be in 2~3 mm/day
![image](https://user-images.githubusercontent.com/97998383/231802034-e8d2cb4b-92e7-44db-8116-6dac9c076399.png)


I really do not konw what happens inside the model. Any help would be greatly appreciated!

**Reply by LianGong98** [reply]
Great suggestion! Next ,I will check my GPP. Thank you so much for your assistance!

**Reply by rgknox** [reply]
Moved my response to the newly created discussion on this
